internal/expr/query.go

```diff
@@ -76,67 +75,95 @@ func (pe *PreparedExpr) Query(args ...any) (ce *QueryExpr, err error) {
 		}
-		named := sql.Named("sqlair_"+strconv.Itoa(i), v.FieldByIndex(in.index).Interface())
-		qargs = append(qargs, named)
 	}
-
 	return &QueryExpr{outputs: pe.outputs, sql: pe.sql, args: qargs}, nil
 }
 
+// MapDecodeInfo stores a map and the results to go in it.
+// Once values have been scanned into the values slice Populate() must be called
+// to set the values in the map.
+type MapDecodeInfo struct {
+	m      reflect.Value
+	keys   []string
+	values []reflect.Value
+}
+
 // ScanArgs returns list of pointers to the struct fields that are listed in qe.outputs.
-// All the structs mentioned in the query must be in outputArgs.
-// All outputArgs must be structs.
-func (qe *QueryExpr) ScanArgs(columns []string, outputArgs []any) ([]any, error) {
+// All the structs and maps mentioned in the query must be in outputArgs.
+func (qe *QueryExpr) ScanArgs(columns []string, outputArgs []any) ([]any, []*MapDecodeInfo, error) {
+	var typesInQuery = []string{}
+	var inQuery = make(map[reflect.Type]bool)
+	for _, typeMember := range qe.outputs {
+		if ok := inQuery[typeMember.outerType()]; !ok {
+			inQuery[typeMember.outerType()] = true
+			typesInQuery = append(typesInQuery, typeMember.outerType().Name())
+		}
+	}
+
+	var mapDecodeInfos = make(map[reflect.Type]*MapDecodeInfo)
+	var typeDest = make(map[reflect.Type]reflect.Value)
 	outputVals := []reflect.Value{}
 	for _, outputArg := range outputArgs {
 		if outputArg == nil {
-			return nil, fmt.Errorf("need pointer to struct, got nil")
+			return nil, nil, fmt.Errorf("need map or pointer to struct, got nil")
 		}
 		outputVal := reflect.ValueOf(outputArg)
-		if outputVal.Kind() != reflect.Pointer {
-			return nil, fmt.Errorf("need pointer to struct, got %s", outputVal.Kind())
+		k := outputVal.Kind()
+		if k != reflect.Map {
+			if k != reflect.Pointer {
+				return nil, nil, fmt.Errorf("need map or pointer to struct, got %s", k)
+			}
+			if outputVal.IsNil() {
+				return nil, nil, fmt.Errorf("got nil pointer")
+			}
+			outputVal = outputVal.Elem()
+			k = outputVal.Kind()
+			if k != reflect.Struct && k != reflect.Map {
+				return nil, nil, fmt.Errorf("need map or pointer to struct, got pointer to %s", k)
+			}
 		}
-		if outputVal.IsNil() {
-			return nil, fmt.Errorf("got nil pointer")
+		if !inQuery[outputVal.Type()] {
+			return nil, nil, fmt.Errorf("type %q does not appear in query, have: %s", outputVal.Type().Name(), strings.Join(typesInQuery, ", "))
 		}
-		outputVal = reflect.Indirect(outputVal)
-		if outputVal.Kind() != reflect.Struct {
-			return nil, fmt.Errorf("need pointer to struct, got pointer to %s", outputVal.Kind())
+		if outputVal.Kind() == reflect.Map {
+			mapDecodeInfos[outputVal.Type()] = &MapDecodeInfo{m: outputVal}
+		} else {
+			if _, ok := typeDest[outputVal.Type()]; ok {
+				return nil, nil, fmt.Errorf("type %q provided more than once, rename one of them", outputVal.Type().Name())
+			}
+			typeDest[outputVal.Type()] = outputVal
```

Comment:
For the case of the first branch
i.e.:
```Go
if k != reflect.Map {
	if k != reflect.Pointer {
		...
```
This makes a special case for maps that have been passed directly in (not as a reference). With Maps this causes no problems and it would be an artificial constraint to only allow them to be passed as pointers. We also allow them to be passed as pointers in which case the branch code will be triggered and the error checks will be carried out (since k will be `reflect.Pointer`)

For the branch in that block I have removed the special case for map. You're right that that error seems equally valid for both.